internal/containers/provider/docker.go

```diff
@@ -42,49 +56,73 @@ func NewDocker(name string, req DockerRequest) (*DockerProvider, error) {
 		name: name,
 		req:  req,
 		cli:  cli,
+		labels: map[string]string{
+			"imagetest": "true",
+		},
 	}, nil
 }
 
-// Start implements Provider.
-func (p *DockerProvider) Start(ctx context.Context) error {
-	if err := p.pull(ctx, p.req.Image); err != nil {
-		return fmt.Errorf("pulling image: %w", err)
+// CreateNetwork creates a user defined bridge network with the given name only
+// if it doesn't exist.
+func (p *DockerProvider) CreateNetwork(ctx context.Context, name string) (string, error) {
+	res, err := p.cli.NetworkInspect(ctx, name, types.NetworkInspectOptions{})
+	if err == nil {
+		return res.ID, nil
 	}
 
-	mode, err := p.cli.NetworkCreate(ctx, p.name, types.NetworkCreate{
-		Attachable: true,
+	mode, err := p.cli.NetworkCreate(ctx, name, types.NetworkCreate{
+		Attachable: false,
 		Driver:     "bridge",
+		Labels:     p.labels,
+		Options: map[string]string{
+			// These are defaults on most daemons, but explicitly set these to ensure
+			// consistency
+			"com.docker.network.bridge.enable_ip_masquerade": "true",
+			"com.docker.network.driver.mtu":                  "65535",
+		},
 	})
+	if err != nil {
+		return "", err
+	}
+	return mode.ID, nil
+}
+
+// Start implements Provider.
+func (p *DockerProvider) Start(ctx context.Context) error {
+	nw, err := p.CreateNetwork(ctx, DockerDefaultNetworkName)
 	if err != nil {
 		return fmt.Errorf("creating network: %w", err)
 	}
 
 	config := &container.Config{
-		Image:        p.req.Image,
+		Image:        p.req.Ref.Name(),
 		User:         p.req.User,
 		Env:          p.req.Env.ToSlice(),
 		Entrypoint:   p.req.Entrypoint,
 		Cmd:          p.req.Cmd,
 		AttachStdout: true,
 		AttachStderr: true,
-		Labels: map[string]string{
-			"imagetest": "true",
-		},
+		Labels:       p.labels,
 	}
 
 	hostConfig := &container.HostConfig{
-		NetworkMode: container.NetworkMode(mode.ID),
+		NetworkMode: container.NetworkMode(nw),
 		Mounts:      p.req.Mounts,
 		Privileged:  p.req.Privileged,
 		RestartPolicy: container.RestartPolicy{
-			Name: "no",
+			Name:              "on-failure",
+			MaximumRetryCount: 1,
```

Comment:
So the container will now try to restart if it enters a failed state? Any reason for this?